src/repart/repart.c

```diff
@@ -5824,76 +5824,80 @@ static int add_subvolume_path(const char *path, Set **subvolumes) {
         if (r < 0)
                 return log_error_errno(r, "Failed to stat source file '%s/%s': %m", strempty(arg_copy_source), path);
 
-        r = set_ensure_consume(subvolumes, &inode_hash_ops, TAKE_PTR(st));
+        r = hashmap_ensure_put(subvolumes, &inode_hash_ops, TAKE_PTR(st), UINT_TO_PTR(flags));
         if (r < 0)
                 return log_oom();
+        if (r == 0)
+                log_debug("Found duplicate subvolume path %s for the same partition", path);
 
         return 0;
 }
 
-static int make_subvolumes_strv(const Partition *p, char ***ret) {
-        _cleanup_strv_free_ char **subvolumes = NULL;
+static int make_subvolumes_hashmap(const Partition *p, Hashmap **ret) {
+        _cleanup_hashmap_free_ Hashmap *hashmap = NULL;
+        _cleanup_free_ char *path = NULL;
         Subvolume *subvolume;
         int r;
 
         assert(p);
         assert(ret);
 
-        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes)
-                if (strv_extend(&subvolumes, subvolume->path) < 0)
+        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes) {
+                path = strdup(subvolume->path);
+                if (!path)
                         return log_oom();
 
-        if (p->suppressing) {
-                char **suppressing;
-
-                r = make_subvolumes_strv(p->suppressing, &suppressing);
-                if (r < 0)
-                        return r;
-
-                r = strv_extend_strv_consume(&subvolumes, suppressing, /* filter_duplicates= */ true);
+                r = hashmap_ensure_put(&hashmap, &string_hash_ops_free, path, UINT_TO_PTR(subvolume->flags));
                 if (r < 0)
                         return log_oom();
+                if (r == 0)
+                        log_debug("Found duplicate subvolume path %s for the same partition", path);
+                TAKE_PTR(path);
         }
 
-        *ret = TAKE_PTR(subvolumes);
+        if (p->suppressing)
+                if (make_subvolumes_hashmap(p->suppressing, &hashmap) < 0)
```

Comment:
You can't pass the existing hashmap to this function, please create a new hashmap `suppressing`, call this function and then merge all its entries into the first one, similar to what the current strv code does